### PR TITLE
Utility bill calculation measure, phase 2

### PR DESCRIPTION
**Pull request by joseph-robertson**
## Pull Request Description

Addresses #920. Adds detailed calculations for tiered, time-of-use, or real-time pricing electric rates.

## Checklist

PR Author: Check these when they're done. Not all may apply. ~~strikethrough~~ and check any that do not apply. 

PR Reviewer: Verify each has been completed.

- [x] Schematron validator (`EPvalidator.xml`) has been updated
- [x] Sample files have been added/updated (via `tasks.rb`)
- [x] Unit tests have been added/updated (e.g., `HPXMLtoOpenStudio/tests`)
- [x] Documentation has been updated
- [x] Changelog has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [ ] Checked the code coverage report on CI
- [x] No unexpected changes to simulation results of sample files

### CODE REVIEW

File: ReportUtilityBills/resources/util.rb
Comment (shorowit) [inline ReportUtilityBills/resources/util.rb:144]: Nice. Can you also incorporate skipping rates with [these keywords](https://github.com/NREL/BEopt2/blob/c84d36bfe42a087e5ff6a6cf4756b2fc31ca0460/Modeling/Utilities/get_utility_rates.py#L109-L121)?

**Review by shorowit — COMMENTED** [review]
A few initial comments:

- Lets further downselect tariffs of interest (potentially to include as a compressed archive). Lets remove any tariff with an `enddate`. Let's also remove any tariff with non-zero demand charges. **There's only a handful; I just left them in.** How many tariffs does that put us at? **5,949.**
- Move sample files to tests directory. Update Sample Real-Time Pricing file so that it reflects the updated JSON schema.
- Rename JSON files to "[utility] - [name].json" or "[utility] - [name] (Effective: [startdate]).json". **Can't have special characters in file names.**

File: ReportUtilityBills/resources/util.rb
Comment (shorowit) [inline ReportUtilityBills/resources/util.rb:541]: add "grain" to the list, I found a few more agricultural rates.

File: ReportUtilityBills/resources/util.rb
Comment (shorowit) [inline ReportUtilityBills/resources/util.rb:547]: I looked at the JSONs sorted by file size and found a few that were really small and basically have no utility rate info in them. Can you filter out tariffs that don't have the minimum fields needed to perform the calculation?

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:329]: I'm confused, why do you look for two different fields? The [documentation](https://openei.org/services/doc/rest/util_rates/?version=3#response-fields) only lists `fixedmonthlycharge`, what is this other one?

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:338]: Does this warning prevent the utility bills from being calculated? I would think it should since it's an incomplete calculation when demand charges are excluded.

Can you exclude JSONs from the tar.gz that have these fields?

File: ReportUtilityBills/tests/Sample Tiered Rate.json
Comment (shorowit) [inline ReportUtilityBills/tests/Sample Tiered Rate.json:10]: I'm confused by some of the fields in these sample files. For example, these three fields aren't in the [documentation](https://openei.org/services/doc/rest/util_rates/?version=3#response-fields), so why are they here? We need to make these sample files consistent with version 3 rates from URDB.

Does your code actually use these fields?

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb:329]: I saw `fixedchargefirstmeter` in at least one tariff file.

File: ReportUtilityBills/tests/Sample Tiered Rate.json
Comment (joseph-robertson) [reply ReportUtilityBills/tests/Sample Tiered Rate.json:10]: I believe it uses `flatratebuy` and `flatratefueladj`.

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb:329]: Hmm perhaps usurdb.csv download is version 1? For example, column `fixedchargefirstmeter` exists and is documented. However, version 3's `fixedmonthlycharge` does not exist as a column.

Comment (shorowit) [reply ReportUtilityBills/measure.rb:329]: If we need to support `fixedchargefirstmeter`, the code above is definitely not correct. It's always being treated as a *monthly* fixed charge, but there is a separate `fixedchargeunits` which says whether it's daily/monthly/annual.

File: ReportUtilityBills/tests/Sample Tiered Rate.json
Comment (joseph-robertson) [reply ReportUtilityBills/tests/Sample Tiered Rate.json:10]: I removed `flatratebuy`, `flatratefueladj`, and `fixedchargeunits` from the "Sample *.json" test files. I updated the `self.detailed_electric` method to no longer use `flatratebuy` or `flatratefueladj`.

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb:329]: I updated the `process_usurdb` method to map to `fixedmonthlycharge`, `minmonthlycharge`, and `annualmincharge` fields based on values in `fixedchargeunits` and `minchargeunits`.

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb:338]: Yes, it does prevent the utility bill (for that scenario) from being calculated.

I _could_. Although keep in mind that we have a unit test that operates on a tariff with demand charges.

Comment (shorowit) [reply ReportUtilityBills/measure.rb:338]: 👍 that it prevents the utility bill from being calculated and 👍 that you have a unit test.

I'd prefer that the tar.gz exclude these JSONs. It will be confusing to a BEopt user if we don't.

File: ReportUtilityBills/measure.rb
Comment (shorowit) [reply ReportUtilityBills/measure.rb:329]: Sounds good. Do our calculations operate on all three (fixedmonthlycharge, minmonthlycharge, and annualmincharge) of those fields? BEopt has inputs for the first two, but not the latter... maybe I should add it.

File: ReportUtilityBills/resources/util.rb
Comment (shorowit) [reply ReportUtilityBills/resources/util.rb:547]: Maybe we can add a unit test around one of these files too.

Comment (joseph-robertson) [reply ReportUtilityBills/resources/util.rb:547]: I filtered out those without `energyweekdayschedule`, `energyweekendschedule`, or `energyratestructure`.

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb]: I think annualmincharge is used for "California-style calculations". See the "TODO" items in `util.rb`. I posted this as a question at the top of this PR.

File: ReportUtilityBills/resources/util.rb
Comment (joseph-robertson) [reply ReportUtilityBills/resources/util.rb:562]: I added `test_warning_missing_required_fields`.

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb:337]: I'll exclude packaging rates with any of the following fields: `demandweekdayschedule`, `demandweekendschedule`, `demandratestructure`, or `flatdemandstructure`.

**Comment by shorowit** [comment]
- [ ] Identify differences between OpenEI v3 and v7; if not very different, update to v7. If significantly different, stick with v3.
- [ ] Specify what version of OpenEI files we expect in the measure arg (and documentation?).

File: ReportUtilityBills/tests/utility_bills_test.rb
Comment (shorowit) [inline ReportUtilityBills/tests/utility_bills_test.rb:1403]: Why can't you do this once up front?

Comment (joseph-robertson) [reply ReportUtilityBills/tests/utility_bills_test.rb:1403]: I am. I put a "skip" in there to prevent it from running on CI.

Comment (joseph-robertson) [reply ReportUtilityBills/tests/utility_bills_test.rb:1403]: Oh, sorry. I re-read. You mean literally the `_load_timeseries` line?

Comment (shorowit) [reply ReportUtilityBills/tests/utility_bills_test.rb:1403]: Yes. Should only need to call it a single time in the method, no?

File: ReportUtilityBills/resources/util.rb
Comment (shorowit) [inline ReportUtilityBills/resources/util.rb:716]: We should find out the answer to this before merging. I noticed that there are ~1000 additional JSONs now, so it's possible that some of them were being filtered out by this previously (and should continue to be filtered out).

File: ReportUtilityBills/resources/util.rb
Comment (shorowit) [inline ReportUtilityBills/resources/util.rb:659]: Let's see if we can find out what's happening here too.

File: BuildResidentialHPXML/measure.rb
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3574]: I don't like the idea of simply ignoring any file paths we don't find. Also, we don't do this for any other similar input in this measure (schedules_filepaths, emissions_electricity_values_or_filepaths, etc.), so we should be consistent.

I would propose just passing the bills_electricity_filepath to the HPXML file without checking anything, and let the downstream HPXMLtoOpenStudio measure take care of checking the path. It looks like there are a few code changes related to making the `hpxml_path` available in this method, so that all could be reverted.

File: HPXMLtoOpenStudio/resources/hpxml_schematron/EPvalidator.xml
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hpxml_schematron/EPvalidator.xml:116]: Isn't there a lot of redundancy here? Seems like all that's needed is:

```suggestion
      <sch:assert role='ERROR' test='count(h:FixedCharge) + count(h:TariffFilePath) &lt;= 1'>Expected 0 or 1 element(s) for xpath: FixedCharge | TariffFilePath</sch:assert>
      <sch:assert role='ERROR' test='count(h:MarginalRate) + count(h:TariffFilePath) &lt;= 1'>Expected 0 or 1 element(s) for xpath: MarginalRate | TariffFilePath</sch:assert>
```

File: HPXMLtoOpenStudio/resources/hpxml_defaults.rb
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hpxml_defaults.rb:287]: Can you update the tests in test_defaults.rb to demonstrate that when there's a tariff file provided, the defaults below are not applied?

File: docs/source/workflow_inputs.rst
Comment (shorowit) [inline docs/source/workflow_inputs.rst:323]: We should probably move away from the simple vs detailed terminology. Yes, the tariff JSON files can be detailed, but they can be simple flat rates too. I will plan to address this.

File: ReportUtilityBills/tests/Sample Flat Rate Min Annual Charge.json
Comment (shorowit) [inline ReportUtilityBills/tests/Sample Flat Rate Min Annual Charge.json:10]: I'm debating whether to make the min charge values (annual and monthly) more realistic. This is a note to self.

File: ReportUtilityBills/tests/utility_bills_test.rb
Comment (shorowit) [inline ReportUtilityBills/tests/utility_bills_test.rb:516]: I don't love that the minimum charges increase the "Marginal" values. Ideally we would rename "Fixed" to "Charges & Fees" or something similar, and have the minimum charges apply to this value instead. (Fixed Charges and Grid Connection Fees already apply to this value.) On the fence as to whether to do this. This is a note to self.

**Review by shorowit — COMMENTED** [review]
This is looking good. I made a couple comments for @joseph-robertson to address and a couple comments for me to think more about.

File: docs/source/workflow_inputs.rst
Comment (shorowit) [reply docs/source/workflow_inputs.rst:323]: On second thought, it's pervasively used everywhere so I will just attempt to clarify what simple vs detailed means, as opposed to making lots of changes.

File: ReportUtilityBills/tests/utility_bills_test.rb
Comment (shorowit) [reply ReportUtilityBills/tests/utility_bills_test.rb:516]: We should really also use "Energy" instead of "Marginal", as is the case in BEopt.

**Review by shorowit — APPROVED** [review]
Thanks, @joseph-robertson! This will merge in if the CI comes back green.

**Pull request merged**
